### PR TITLE
Add 051 release demo notebook

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
CC @gmarkall for CUDA link in.

CC @luk-f-a @StanczakDominik was wondering if either of you would like to do an initial review? No obligation/pressure :) Thanks!

**Comment by StanczakDominik** [comment]
Sure, I'll take a look today :)

**Comment by luk-f-a** [comment]
yes, no problem, I'll have a look today.

**Comment by gmarkall** [comment]
The wording relating to the CUDA notebook looks good. When I click the link to the notebook on Binder, I get a "400: Bad Request" error though - is that to be expected at this stage prior to the merge of #21?

File: notebooks/Numba_051_release_demo.ipynb [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+    "    # By default, the proxy type does not reflect the attributes or\n",
+    "    # methods to the Python side. It is up to users to define\n",
+    "    # these. (This may be automated in the future.)\n",
+    "\n",
+    "    @property\n",
+    "    def kind(self):\n",
+    "        # To access a field, we can define a function that simply\n",
+    "        # return the field in jit-code.\n",
+    "        # The definition is shown later.\n",
+    "        return Fruit_get_kind(self)\n",
```
Comment (luk-f-a) [inline notebooks/Numba_051_release_demo.ipynb:428]: writing this as the usual `return self.kind` crashes Python. Is this a bug? If not, a note highlighting that it must be done via an `njit` function would be helpful.

File: notebooks/Numba_051_release_demo.ipynb [hunk 2]
```diff
@@ -0,0 +1,564 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Numba 0.51.0 Release Demo\n",
+    "=======================\n",
+    "\n",
+    "This notebook contains a demonstration of new features present in the 0.51.0 release of Numba. Whilst release notes are produced as part of the [`CHANGE_LOG`](https://github.com/numba/numba/blob/release0.51/CHANGE_LOG), there's nothing like seeing code in action!\n",
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply notebooks/Numba_051_release_demo.ipynb:10]: Thanks for highlighting this, I think `nbsphinx` could well be useful here, wonder what the complexity of using it cross repo is? The notebooks (and in fact this whole repo) are wired up to binder here https://mybinder.org/v2/gh/numba/numba-examples/master and in release announcements etc the live notebooks are referenced, so it's pre-action or action depending on how it was accessed! :)

For reference, the 0.50 release announcement is here https://numba.discourse.group/t/numba-0-50-0-and-llvmlite-0-33-0-released/19 with binder link https://mybinder.org/v2/gh/numba/numba-examples/master?filepath=notebooks%2FNumba_050_release_demo.ipynb

**Comment by stuartarchibald** [comment]
> The wording relating to the CUDA notebook looks good. When I click the link to the notebook on Binder, I get a "400: Bad Request" error though - is that to be expected at this stage prior to the merge of #21?

@gmarkall  Thanks, and yes, needs to be on `master/` I think for binder to pick it up.

File: notebooks/Numba_051_release_demo.ipynb [hunk 3]
```diff
@@ -0,0 +1,564 @@
...
+    "except errors.TypingError as e:\n",
+    "    # CANNOT MUTATE A LITERAL LIST!\n",
+    "    assert \"Cannot mutate a literal list\" in str(e)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Heterogeneously typed lists also carry their type information, including literal values, such that it's possible to dispatch based on their value types"
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply notebooks/Numba_051_release_demo.ipynb:124]: Just from jitted functions, there's no unboxing for heterogeneous lists.

[the remark this answers is not in the document]
Comment (stuartarchibald) [reply notebooks/Numba_051_release_demo.ipynb:124]: Added a note in the notebook about that.

File: notebooks/Numba_051_release_demo.ipynb [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# User defined mutable pass-by-ref structures\n",
+    "\n",
+    "A very common question from users is:\n",
+    "\n",
+    "> What can I use as a mutable structure that's also pass-by-reference?\n",
+    "\n",
+    "the answer is the new ``StructRef`` type (warning: this is experimental!), [documentation is here](http://numba.pydata.org/numba-doc/dev/extending/high-level.html#implementing-mutable-structures)."
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply notebooks/Numba_051_release_demo.ipynb:382]: Undetermined at present! It's all experimental.

File: notebooks/Numba_051_release_demo.ipynb (hunk 1 above)
Comment (stuartarchibald) [reply notebooks/Numba_051_release_demo.ipynb:428]: The function is required to act as an accessor to the instance of the representation in the compiled code.

**Comment by stuartarchibald** [comment]
Thanks for the review @StanczakDominik @luk-f-a, think I've addressed all your comments, please ping if not. I also had to update the notebook a bit with respect to last minute bug fixes too!

File: notebooks/Numba_051_release_demo.ipynb [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+    "    # a list of type [literal intp, unicode string, NumPy 1d array of float64]\n",
+    "    x = [1, 'a', np.zeros(5)]\n",
+    "    \n",
+    "    x.append(2j) # illegal mutation\n",
+    "\n",
+    "try:\n",
+    "    mixed_type_list_error()\n",
+    "except errors.TypingError as e:\n",
+    "    # CANNOT MUTATE A LITERAL LIST!\n",
+    "    assert \"Cannot mutate a literal list\" in str(e)"
```
[the remark this answers is not in the document]
Comment (sklam) [reply notebooks/Numba_051_release_demo.ipynb:118]: a print was added

File: notebooks/Numba_051_release_demo.ipynb [hunk 6]
```diff
@@ -0,0 +1,564 @@
...
+    "    print(result_b)\n",
+    "\n",
+    "config_example(np.arange(10.))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Initial value capturing\n",
```
[the remark this answers is not in the document]
Comment (sklam) [reply notebooks/Numba_051_release_demo.ipynb:241]: A note was added below

**Comment by sklam** [comment]
Looks like everything is addressed. Merging.

**Pull request merged**